R/serializer.R

```diff
@@ -324,20 +356,173 @@ serializer_xml <- function() {
 
 
 
+#' Hooks and serializer object
+#'
+#' This method allows serializers to return both hooks and a serializer.
+#' This is useful for graphics device serializers which need a `preexec` and `postexec` hook to caputre the graphics output.
+#' @param hooks Hooks to be supplied directly to corresponding [PlumberEndpoint] `$registerHooks()` method
+#' @param serializer Serializer method to be used.  This method should already have it's initialization arugments applied.
+#' @examples
+#' # The definition of `serializer_device` returns
+#' # * `preexec`, `postexec` hooks
+#' # * a `serializer_content_type` serializer
+#' print(serializer_device)
+hooks_and_serializer <- function(hooks, serializer) {
+  structure(
+    list(
+      hooks = hooks,
+      serializer = serializer
+    ),
+    class = "plumber_hooks_and_serializer"
+  )
+}
+
+self_set_serializer <- function(self, serializer) {
+  if (inherits(serializer, "plumber_hooks_and_serializer")) {
+    self$serializer <- serializer$serializer
+    self$registerHooks(serializer$hooks)
+  } else {
+    self$serializer <- serializer
+  }
+  invisible(self)
+}
+
+
+#' @describeIn serializers Helper method to create graphics device serializers, such as [serializer_png()]. See also: [hooks_and_serializer()]
+#' @param dev_on Function to turn on the graphics device.
+#' The graphics device `dev_on` function will receive any arguments supplied to the serializer in addition to `filename`.
+#' `filename` points to the temporary file name that should be used when saving content.
+#' @param dev_off Function to turn off the grahpics device. Defaults to [grDevices::dev.off()]
+#' @export
+serializer_device <- function(type, dev_on, dev_off = grDevices::dev.off) {
```

Comment:
```suggestion
serializer_device <- function(type, dev_on = grDevices::dev.new, dev_off = grDevices::dev.off) {
```

Reply:
I am hesitant on this one as I'd like to make that a required argument. 

I know `dev.new()` opens the default graphics device, but not all devices share the argument of `filename`. Some (ex: `quartz()` and `pdf()`) accept `file`.  `quartz()` is the default on macOS.

Should I have the calling function look for argument of `file` or `filename` in the `dev_on()` function and use the appropriate one?